### PR TITLE
Carmen checkpoint

### PR DESCRIPTION
**Pull request by cabrador**
## Description

This PR adds ability to set how often will Carmen create checkpoints

- [x] #1146 is merged

## Type of change

Please delete options that are not relevant.

- [x] New feature (non-breaking change which adds functionality)

### CODE REVIEW

File: cmd/aida-vm-sdb/run_substate.go [hunk 1]
```diff
@@ -94,7 +94,7 @@ func runSubstates(cfg *utils.Config, provider executor.Provider[txcontext.TxCont
 		profiler.MakeVirtualMachineStatisticsPrinter[txcontext.TxContext](cfg),
 		logger.MakeProgressLogger[txcontext.TxContext](cfg, 15*time.Second),
 		logger.MakeErrorLogger[txcontext.TxContext](cfg),
-		tracker.MakeBlockProgressTracker(cfg, 100_000),
+		tracker.MakeBlockProgressTracker(cfg, 100),
```
Comment (wsodsong) [inline cmd/aida-vm-sdb/run_substate.go]: This prints progress every 100 block. Is it for debugging or you want to increase the frequency?

File: executor/extension/statedb/live_db_block_checker.go [hunk 2]
```diff
@@ -85,6 +85,11 @@ func (c *liveDbBlockChecker[T]) PreRun(executor.State[T], *executor.Context) err
 			return fmt.Errorf("if using existing live-db with vm-sdb first block needs to be last block of live-db + 1, in your case %v", lastBlock+1)
 		}
 	} else if lastBlock >= c.cfg.First {
+		// Ignore if run was not finished - db might have been healed so block alignment would not work
```
Comment (wsodsong) [inline executor/extension/statedb/live_db_block_checker.go:88]: If a src db is given, then I don't think it should perform priming.

File: utils/default_config.go [hunk 3]
```diff
@@ -40,6 +40,8 @@ func createConfigFromFlags(ctx *cli.Context) *Config {
 		CPUProfilePerInterval:  getFlagValue(ctx, CpuProfilePerIntervalFlag).(bool),
 		Cache:                  getFlagValue(ctx, CacheFlag).(int),
 		CarmenSchema:           getFlagValue(ctx, CarmenSchemaFlag).(int),
+		CarmenCheckpointInterval: getFlagValue(ctx, CarmenCheckpointInterval).(int),
+		CarmenCheckpointPeriod:   getFlagValue(ctx, CarmenCheckpointPeriod).(int),
```
Comment (wsodsong) [inline utils/default_config.go]: nit: keep alphabetical order 
```suggestion
		CarmenCheckpointInterval: getFlagValue(ctx, CarmenCheckpointInterval).(int),
		CarmenCheckpointPeriod:   getFlagValue(ctx, CarmenCheckpointPeriod).(int),
		CarmenSchema:                   getFlagValue(ctx, CarmenSchemaFlag).(int),
```

File: utils/statedb.go [hunk 4]
```diff
@@ -201,20 +201,41 @@ func makeNewStateDB(cfg *Config) (state.StateDB, string, error) {
...
+			variant,
+			rootHash,
+			cfg.ArchiveMode,
+			state.NewChainConduit(cfg.ChainID == EthereumChainID, cfg.ChainCfg),
+		)
 	case "carmen":
 		// Disable archive if not enabled.
 		if !cfg.ArchiveMode {
-			archiveVariant = "none"
+			cfg.ArchiveVariant = "none"
```
Comment (wsodsong) [inline utils/statedb.go]: `cfg.ArchiveVariant` is not passed to `MakeCarmenStateDB` at line 231.

File: utils/statedb.go [hunk 5]
```diff
@@ -201,20 +201,41 @@ func makeNewStateDB(cfg *Config) (state.StateDB, string, error) {
...
+			carmenSchema,
+			archiveVariant,
+			cfg.CarmenNodeCacheSize,
+			cfg.CarmenNodeCacheSize,
+			cfg.CarmenCheckpointInterval,
+			cfg.CarmenCheckpointPeriod,
+		)
 	}
-	return nil, fmt.Errorf("unknown Db implementation: %v", impl)
+	return nil, fmt.Errorf("unknown Db implementation: %v", cfg.DbImpl)
```
Comment (wsodsong) [inline utils/statedb.go]: I think we should keep `impl`. `impl`, which is passed as an argument, is the target implementation. This matters when preparing a shadow db. `cfg.impl` only refer to db implementation of the primary  db.

File: executor/extension/statedb/live_db_block_checker.go (hunk 2 above)
Comment (cabrador) [reply executor/extension/statedb/live_db_block_checker.go:88]: I remember @matejmode added feature allowing priming of existing database up to some block.

File: utils/statedb.go (hunk 5 above)
Comment (cabrador) [reply utils/statedb.go]: Good catch. Thanks!

File: executor/extension/statedb/live_db_block_checker.go (hunk 2 above)
Comment (wsodsong) [reply executor/extension/statedb/live_db_block_checker.go:88]: That is true. The issue is that it relies on a block number from statedb info. Do you have a way to generate statedb info?

File: executor/extension/statedb/state_db_manager.go [hunk 6]
```diff
@@ -90,6 +91,11 @@ func (m *stateDbManager[T]) PreRun(_ executor.State[T], ctx *executor.Context) e
 	if !m.cfg.KeepDb && !m.cfg.StateDbSrcDirectAccess {
 		m.log.Warningf("--keep-db is not used. Directory %v with DB will be removed at the end of this run.", ctx.StateDbPath)
 	}
+
+	if err := utils.WriteStateDbInfo(ctx.StateDbPath, m.cfg, 0, gc.Hash{}, false); err != nil {
```
Comment (wsodsong) [inline executor/extension/statedb/state_db_manager.go:95]: Is the intention here is to first write statedb info before start and mark `hasFinished` as false; then overwrite the file with `hasFinished` as true on completion?

File: executor/extension/statedb/live_db_block_checker.go (hunk 2 above)
Comment (petr-hanzl) [reply executor/extension/statedb/live_db_block_checker.go:88]: I was thinking about it. I havent found any way

Comment (cabrador) [reply executor/extension/statedb/live_db_block_checker.go:88]: Well I write it before the run and mark it as `unfinished` as mentioned bellow

File: executor/extension/statedb/state_db_manager.go (hunk 6 above)
Comment (cabrador) [reply executor/extension/statedb/state_db_manager.go:95]: Thats right.

File: utils/flags.go [hunk 7]
```diff
@@ -54,6 +54,16 @@ var (
 		Usage: "sets the balance range of the stochastic simulation",
 		Value: 1000000,
 	}
+	CarmenCheckpointInterval = cli.IntFlag{
+		Name:  "carmen-cp-interval",
```
Comment (wsodsong) [inline utils/flags.go]: nit: I would spell it out `carmen-checkpoint-interval`

File: utils/flags.go [hunk 8]
```diff
@@ -54,6 +54,16 @@ var (
...
+		Usage: "defines how often (in blocks) will Carmen create checkpoints",
+		Value: 0,
+	}
+	CarmenCheckpointPeriod = cli.IntFlag{
+		Name:  "carmen-cp-period",
```
Comment (wsodsong) [inline utils/flags.go]: nit: here as well `carmen-checkpoint-period`

**Review by wsodsong — DISMISSED** [review]
Just minor comments of cli argument names, everything else looks good.

**Review by wsodsong — DISMISSED** [review]

**Review by wsodsong — DISMISSED** [review]

**Review by wsodsong — APPROVED** [review]

**Pull request merged**